### DISCUSSION TITLE
New livecd version news update

### DISCUSSION THREAD

**Discussion by immolo** [frno7/mipsr5900el-gentoo-linux-gnu#19] [General]
Just an update for anyone interested on what is going on.

I have been doing tests using the new mips 23.0 musl profiles with a snapshot of musl 1.2.4 and LTO to reduce RAM usage further and things were going smoothly until the LFS changes upstream [see bug](https://bugs.gentoo.org/900871) which means this is hit a brick until someone solves it. 

Also for fun I tried an llvm only build but from early testing it seems the r5900 isn't a supported CPU currently so if anyone wants to fix that we could reduce the memory even further if ti follows the trends of other systems I have tested this on.

**Comment by frno7** [comment]
Nice! Does LLVM emit proper MIPS II, perchance? The R5900 is happy with MIPS I and II, provided its [short loop erratum](https://github.com/frno7/linux/wiki/R5900-short-loop-erratum) is handled. The GNU assembler can do that, if it’s possible to have LLVM emit assembler files for GAS?

**Reply by immolo** [reply]
This is pushing way past my knowledge so most of this is just brain dumping what I have found so please do correct or highlight where I've missed.

From the manual it seems you can use GAS:
```
Inline assembly

In general, Clang is highly compatible with the GCC inline assembly extensions, allowing the same set of constraints, modifiers and operands as GCC inline assembly.

On targets that use the integrated assembler (such as most X86 targets), inline assembly is run through the integrated assembler instead of your system assembler (which is most commonly "gas", the GNU assembler). The LLVM integrated assembler is extremely compatible with GAS, but there are a couple of minor places where it is more picky, particularly due to outright GAS bugs.

One specific example is that the assembler rejects ambiguous X86 instructions that don't have suffixes. For example:

  asm("add %al, (%rax)");
  asm("addw $4, (%rax)");
  asm("add $4, (%rax)");

Both clang and GAS accept the first instruction: because the first instruction uses the 8-bit %al register as an operand, it is clear that it is an 8-bit add. The second instruction is accepted by both because the "w" suffix indicates that it is a 16-bit add. The last instruction is accepted by GAS even though there is nothing that specifies the size of the instruction (and the assembler randomly picks a 32-bit add). Because it is ambiguous, Clang rejects the instruction with this error message:

<inline asm>:3:1: error: ambiguous instructions require an explicit suffix (could be 'addb', 'addw', 'addl', or 'addq')
add $4, (%rax)
^

To fix this compatibility issue, add an explicit suffix to the instruction: this makes your code more clear and is compatible with both GCC and Clang.
```

I'd need to learn how to do this though so and hints would be helpful, although I just found this will looking at GAS related LLVM for the r5900 https://lists.llvm.org/pipermail/llvm-dev/2018-September/125800.html so this might give me some pointers.

Looks like the way supported mips cpus are handled in LLVM16+ might make life a little easier so it could be worth trying this as the next base [as seen here](https://clang.llvm.org/doxygen/Basic_2Targets_2Mips_8cpp_source.htmll) but to answer your question I can't find the 6 CPUs which were supported originally now so I'll get back to you if I can find those again.

I should have added previously that I've also tried the latest snapshot of GCC-13 to make sure this still works and all early signs show there will be no issues moving forward on this part.

**Reply by frno7** [reply]
So given for example `hullo.c` as in

```c
#include <stdio.h>
int main()
{
	puts("Hullo");
	return 0;
}
```

we can do

```
% mipsr5900el-unknown-linux-musl-gcc -S -o hullo.S hullo.c
% mipsr5900el-unknown-linux-musl-as -o hullo.o hullo.S
% mipsr5900el-unknown-linux-musl-gcc -static -o hullo hullo.o
% qemu-mipsr5900el hullo
Hullo
```

Now, in the first command one could try to exchange GCC for Clang, to have Clang emit `hullo.S` instead, and proceed with the rest of the commands as before. Does that work?

> Looks like the way supported mips cpus are handled in LLVM16+ might make life a little easier so it could be worth trying this as the next base [as seen here](https://clang.llvm.org/doxygen/Basic_2Targets_2Mips_8cpp_source.html) but to answer your question I can't find the 6 CPUs which were supported originally now so I'll get back to you if I can find those again.

Both `mips1` and `mips2` are valid LLVM targets, as it seems!

```c
static constexpr llvm::StringLiteral ValidCPUNames[] = {
    {"mips1"},  {"mips2"},    {"mips3"},    {"mips4"},    {"mips5"},
    {"mips32"}, {"mips32r2"}, {"mips32r3"}, {"mips32r5"}, {"mips32r6"},
    {"mips64"}, {"mips64r2"}, {"mips64r3"}, {"mips64r5"}, {"mips64r6"},
    {"octeon"}, {"octeon+"}, {"p5600"}};
```

Looks promising to me. :-)

**Reply by immolo** [reply]
Looks interesting and you have given me some interesting points to look at next.

I'll setup a test env on my CI40 once llvm16 comes out of RC stage to see how it works on a slightly easier target and then we can see if we can brainstorm a fix using that test for PS2 support.

As a side note there is work being done on have crossdev support clang so this could become even easier then the current method of using qemu-mipsel to do any llvm work.

**Reply by frno7** [reply]
Sounds good! It should be _relatively_ easy to port the `-mfix-r5900` GCC/GAS option to Clang as well.

**Reply by immolo** [reply]
So I've got llvm16 working on my ci40 and currently have clang compiling but the good news is llvm16 is supposed to keyworded for the 23.0 profile release so there was already a lot of work that went on behind the scenes it seems :)

Currently though testing is very slow as I keep hitting the qemu-user mips 2gb RAM per thread limit so need to compile on real hardware (currently 8 days in). I know this is your area so if you have ideas on how I can remove this limit to speed up work on this then it would be very helpful to us and possibly the Gentoo releng team as they must hit this too with stage3 building.

**Reply by frno7** [reply]
Does compiling on real hardware somehow work better than user QEMU? What packages, intended for the R5900, demand more than 2 GiB to compile, by the way? I guess it’s the o32 ABI? Doesn’t cross-compiling work for these demanding packages?

**Reply by immolo** [reply]
> Does compiling on real hardware somehow work better than user QEMU? What packages, intended for the R5900, demand more than 2 GiB to compile, by the way? I guess it’s the o32 ABI? Doesn’t cross-compiling work for these demanding packages?

gcc, llvm and clang are the only 3 I've found so far but getting clang to cross compile has been a pain but this method works for the other two. Cross compiling will be an issue for stage3 creation though as catalyst uses qemu-user for it.

**Reply by frno7** [reply]
Hmm, all 32-bit systems have similar problems. The hard limit is 4 GiB (as in a 2^32 address space). I believe some systems at least allow 3 GiB, reserving 1 GiB for the operating system. I don’t know whether accessing the range 2-4 GiB is possible with the MIPS o32 ABI. QEMU has the advantage that it can be patched, to do things slightly differently compared to hardware, in principle.

Compiler optimisation is often very, if not _the_ most, memory consuming part of compilation. Maybe one could dial that down a bit? Use `-O1` rather than `-O2`? There are many [options for optimisations](https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html).

The original [68000](https://en.wikipedia.org/wiki/Motorola_68000_series) processor has a 24-bit address bus, effectively limiting it to 16 MiB of memory. Those were the days. :-)

Github is no longer posting email notifications, apparently. Nowadays I have to click through all repos to see any posts, so I’m having some difficulties following conversations.

**Reply by immolo** [reply]
I missed the notification too.

I'm using the `-Os` flag currently but I'm not how much more bloated that is over `-O1` but maybe something to look into.

From what I understand Malta boards could support over 2GB in o32 but I'm just reading mailing lists from the past so I would class this as interesting rather than fact :)

In related news, GCC13 has a new patch which splits gimple up, which was just dropped into 13.2.1 which should make building GCC stage3s a lot more easier if it works like we expect it to. More infomation - https://www.mail-archive.com/gcc@gcc.gnu.org/msg95346.html

**Reply by frno7** [reply]
Regarding unreliable Git Hub notifications, I’ve found the Git Hub web interface notifications (available at https://github.com/notifications) more reliable _in combination_ with email. Most of the time, notes pop up as both email and web items, but twice is better than none surely? :-)

**Reply by immolo** [reply]
Thanks, I'll check this out.

I'm hoping to be more active the next few months as MIPS stuff is looking good in Gentoo to the point I can share the stuff back here.

**Reply by frno7** [reply]
I haven’t heard anything lately. @immolo?

**Comment by immolo** [comment]
@JohnShiozo @frno7 

(I really need to fix notifications)

Upstream in Gentoo is going through some major changes with turple change for all mips based systems which is has slowed down the official musl release and the gcc-13.2.0 drop in ~arch hasn't helped speed things up. (mipsel2_musl just dropped though so things are moving forward.) I haven't been chasing this up as much as I should though as someone donated me a m68k [PowerBook 190](https://everymac.com/systems/apple/powerbook/specs/mac_powerbook190_66.html) so I got a little side tracked as the Gentoo devs let me learn and commit all the changes so Gentoo can have official [m68k support](https://github.com/gentoo/gentoo/commit/68a23a753439ef5439a5ff847de84b0601b47df0).

Let me fix the lack of updates by creating a new issue tracker with a checklist so you can see the milestones for a release for our community to enjoy and beyond that goals to try and get Gentoo PS2 officially supported with community support as we could both benefit working closer together. Please also feel free to complete any of the tasks on it to help me out and knowing others are interested boosts my attention to spend my limited time on tasks people will use.

@frno7 side point you'll be interested in, you should look into rebasing your qemu patches to 8.0.3 as someone found and fixed an issue in memory usage which everyone thought was normal and worked around for years. I can't find the commit this second but it has fixed so many issues for me that I'm pretty sure it will help you out too even if it's just faster compiling.

**Reply by immolo** [reply]
#21 is up as a rough guide now but I'll clean it up later on when I have more time. Ideas and suggestions are always welcome :)

**Reply by frno7** [reply]
> (I really need to fix notifications)

@immolo, yeah, activating one's [personal Git Hub web notification page](https://github.com/notifications) only takes a couple of clicks, and it works _in combination with email_ too. I’ve lost quite a few Git Hub email notifications, but I no longer worry about it now that this web page provides notification redundancy. :-)

> I haven't been chasing this up as much as I should though as someone donated me a m68k [PowerBook 190](https://everymac.com/systems/apple/powerbook/specs/mac_powerbook190_66.html) so I got a little side tracked as the Gentoo devs let me learn and commit all the changes so Gentoo can have official [m68k support](https://github.com/gentoo/gentoo/commit/68a23a753439ef5439a5ff847de84b0601b47df0).

Happy days. :-) By the way, did you figure out how to avoid memory usage from ballooning during 32-bit compilation?

> Let me fix the lack of updates by creating a new issue tracker with a checklist so you can see the milestones for a release for our community to enjoy and beyond that goals to try and get Gentoo PS2 officially supported with community support as we could both benefit working closer together. Please also feel free to complete any of the tasks on it to help me out and knowing others are interested boosts my attention to spend my limited time on tasks people will use.

Nice!

> @frno7 side point you'll be interested in, you should look into rebasing your qemu patches to 8.0.3 as someone found and fixed an issue in memory usage which everyone thought was normal and worked around for years. I can't find the commit this second but it has fixed so many issues for me that I'm pretty sure it will help you out too even if it's just faster compiling.

The previous QEMU maintainer for MIPS wasn’t happy at all with the R5900 and the o32 ABI, so upgrading became progressively difficult after QEMU 5.2. Can the patch be backported to 5.2?

A year or so after the R5900 QEMU incident the MIPS company went bankrupt, and now it’s apparently into RISC-V instead.

**Comment by immolo** [comment]
> Happy days. :-) By the way, did you figure out how to avoid memory usage from ballooning during 32-bit compilation?

GCC split gimple saving me the hassle and the patch in qemu 8.0.3 has been helping a lot with random errors. I've not had time to look into llvm as of yet.

> The previous QEMU maintainer for MIPS wasn’t happy at all with the R5900 and the o32 ABI, so upgrading became progressively difficult after QEMU 5.2. Can the patch be backported to 5.2?
> 
> A year or so after the R5900 QEMU incident the MIPS company went bankrupt, and now it’s apparently into RISC-V instead.

Very good question, the patch likely could but there has been so many changes to qemu-user that I would be leaning on using the latest version, although do bear in mind I only use qemu for compiling so other usage cases might make more sense on the older versions.

While I'm here, I had some time over the weekend nursing a hangover so got bored waiting for upstream to drop a new stage and have created my own. At the time of writing stage1 is built and stage3 is 1/4 done, I've enabled [LTO](https://wiki.gentoo.org/wiki/LTO) for morbid curiosity oh how low I can get the ram to. Hopefully work eases up a bit that I can get something out soon.

**Reply by frno7** [reply]
> GCC split gimple saving me the hassle and the patch in qemu 8.0.3 has been helping a lot with random errors. I've not had time to look into llvm as of yet.

I suppose changing `-O2` to `-O1` in `COMMON_FLAGS="-O2 -pipe"` in `/etc/portage/make.conf` could reduce compiler memory pressure (and compile faster too). Optimisations won’t be as good, of course.

> While I'm here, I had some time over the weekend nursing a hangover so got bored waiting for upstream to drop a new stage and have created my own. At the time of writing stage1 is built and stage3 is 1/4 done, I've enabled [LTO](https://wiki.gentoo.org/wiki/LTO) for morbid curiosity oh how low I can get the ram to. Hopefully work eases up a bit that I can get something out soon.

Cool. :-)
